src/input/mod.rs

```diff
@@ -305,6 +307,32 @@ impl State {
         self.handle_bind(bind);
     }
 
+    fn apply_keyboard_config<I: InputBackend>(&mut self, event: &I::KeyboardKeyEvent) {
+        let device = event.device();
+
+        let keyboard_config = self
+            .niri
+            .config
+            .borrow()
+            .input
+            .keyboard_named(device.name());
```

Comment:
This `device.name()` allocates a new `String` every time, not a big fan of that... Maybe we should have a cache `HashMap<Device, niri_config::Keyboard>`?

Reply:
i don't think it's currently possible since `Device` supers `PartialEq` which can't be a trait object. please correct me if i'm wrong.

Reply:
ahh can use a `HashMap<input::Device, niri_config::Keyboard>` and just downcast.